storage: add storage version field and adapt old version

liuX10 opened the pull request.

liuX10 committed:
storage: Add storage version number and fix compatibility issues-KVDB.

bug: v/54525

Signed-off-by: liuxiang18 <liuxiang18@xiaomi.com>

gzh-terry commented on service/common/storage.c:
```diff
@@ -44,6 +44,7 @@ typedef struct {
 } key_header_t;
 
 static uv_db_t* storage_handle = NULL;
+const uint8_t bt_prop_version_0_1[VERSION_HEADER_LENGTH] = { 0x33, 0x8B, 0x9E, 0x00, 0x00, 0x01 };
```
`static const uint8_t`

gzh-terry commented on service/common/storage.c:
```diff
@@ -184,6 +185,30 @@ void bt_storage_load_irk_info(void)
+    memcpy(version, bt_prop_version_0_1, VERSION_HEADER_LENGTH);
+}
+
+bool version_match(void* version)
```
`static`, same for others.

gzh-terry replied:
add `bt_storage_xxx` for global functions.
otherwise add `static`

gzh-terry replied:
`bool bt_storage_version_match(const uint8_t* version)`

gzh-terry commented on service/common/storage.c:
```diff
@@ -184,6 +185,30 @@ void bt_storage_load_irk_info(void)
+    return memcmp(bt_prop_version_0_1, version, VERSION_HEADER_LENGTH) == 0;
+}
+
+void bt_property_transform(remote_device_old_properties_t* src_prop, remote_device_properties_t* dst_prop, bool load_uuid)
```
`dst` first, then `src`

gzh-terry commented on service/src/adapter_service.c:
```diff
@@ -304,12 +304,29 @@ static void load_remote_uuids(remote_device_properties_t* remote, bt_device_t* d
     if (data && items) {
         char addr_str[BT_ADDR_STR_LENGTH] = { 0 };
-        remote_device_properties_t* remote = (remote_device_properties_t*)data;
+        is_match = version_match(data);
```
`version_match` returns a `true` if `*data` is exactly the current version.
However, current version can be `bt_prop_version_0_1` or `bt_prop_version_0_2`
Make sure `is_match` has a correct behavior in case of parsing `bt_prop_version_0_1` when current version is `bt_prop_version_0_2`

gzh-terry commented on service/src/adapter_service.c:
```diff
@@ -318,17 +335,23 @@ static void bonded_device_loaded(void* data, uint16_t length, uint16_t items)
             bt_sal_set_bonded_devices(PRIMARY_ADAPTER, remote, 1);
-            remote++;
+            if (!is_match && i < items) {
+                bt_property_transform((remote_device_old_properties_t*)((uint8_t*)data + i * item_len), remote, load_uuid);
```
Caution: `i` starts from 1 now.
See line 329 `for (int i = 1; i <= items; i++) {`

liuX10 replied:
yes! first transform at line 323, Range from 1 to items - 1. 
if items == 1, it would not go this way.

gzh-terry commented on service/src/adapter_service.c:
```diff
@@ -304,12 +304,29 @@ static void load_remote_uuids(remote_device_properties_t* remote, bt_device_t* d
+    remote_device_properties_t* remote_tmp = zalloc(sizeof(remote_device_properties_t));
+
     if (data && items) {
         char addr_str[BT_ADDR_STR_LENGTH] = { 0 };
```
Move to the top

gzh-terry commented on service/src/adapter_service.c:
```diff
@@ -318,17 +335,23 @@ static void bonded_device_loaded(void* data, uint16_t length, uint16_t items)
                 lk[7], lk[8], lk[9], lk[10], lk[11], lk[12], lk[13], lk[14], lk[15]);
             bt_sal_set_bonded_devices(PRIMARY_ADAPTER, remote, 1);
-            remote++;
+            if (!is_match && i < items) {
```
I would suggest to put these in a new method. And have a proper name to illustrate it is used to find/get/fetch the next `remote`

Or at least add some comment to keep the code readable.

liuX10 committed:
storage: Add storage version number and fix compatibility issues.

bug: v/54525

Signed-off-by: liuxiang18 <liuxiang18@xiaomi.com>

liuX10 committed:
storage: fix load error as without 128 type uuid.

bug: v/54525

When a device does not have a 128-bit type UUID, existing code fails to properly initialize the count_uuid128 variable, resulting in subsequent reads of the uninitialized variable that may trigger undefined behavior (including reading a random memory value) and ultimately crash the application.
Signed-off-by: liuxiang18 <liuxiang18@xiaomi.com>

JinxianChe commented on service/common/storage_property.c:
```diff
@@ -54,6 +54,8 @@
 #define BT_KVDB_BLEBOND "persist.bluetooth.blebonded."
 #define BT_KVDB_BLEWHITELIST "persist.bluetooth.whitelist."
 
+static const uint8_t bt_prop_version_0_1[VERSION_HEADER_LENGTH] = { 0x33, 0x8B, 0x9E, 0x00, 0x00, 0x01 };
```
If both storage.c and storage_property.c need to maintain version information, it feels inconvenient. Can they be unified for maintenance?
